### PR TITLE
feat: Simple cart totals calculations

### PR DESCRIPTION
**Pull request by olivermrbl**

### CODE REVIEW

File: packages/cart/integration-tests/__tests__/services/cart-module/index.spec.ts [hunk 1]
```diff
@@ -2332,4 +2333,184 @@ describe("Cart Module Service", () => {
...
+              code: "20%",
+            },
+          ],
+        },
+      ])
+
+      return createdCart
+    }
+
+    const expectedRes = {
```
Comment (olivermrbl) [inline packages/cart/integration-tests/__tests__/services/cart-module/index.spec.ts:2398]: @srindom, would like your eyes on this 😄

**Comment by adrien2p** [comment]
could you add a description and resolve the conflict please :)

File: packages/cart/integration-tests/__tests__/services/cart-module/index.spec.ts [hunk 2]
```diff
@@ -2332,4 +2333,184 @@ describe("Cart Module Service", () => {
       expect(taxLines?.length).toBe(0)
     })
   })
+
+  describe.only("retrieve with totals decoration", () => {
```
Comment (adrien2p) [inline packages/cart/integration-tests/__tests__/services/cart-module/index.spec.ts:2337]: todo: rm only

File: packages/cart/integration-tests/__tests__/services/cart-module/index.spec.ts [hunk 3]
```diff
@@ -2332,4 +2333,184 @@ describe("Cart Module Service", () => {
...
+        },
+      ])
+
+      return createdCart
+    }
+
+    const expectedRes = {
+      subtotal: 600,
+
+      total: 784, // 600 (subtotal) + 116 (shipping) + 112 (item tax) + 16 (shipping tax) - 60 (discount)
```
Comment (adrien2p) [inline packages/cart/integration-tests/__tests__/services/cart-module/index.spec.ts]: q: how do we get 116 for the shipping? wouldn't it be 80 (100-20) and 16 taxes? and the result would be 748

File: packages/cart/integration-tests/__tests__/services/cart-module/index.spec.ts [hunk 4]
```diff
@@ -2332,4 +2333,184 @@ describe("Cart Module Service", () => {
...
+      tax_total: 128,
+
+      item_total: 672,
+      item_subtotal: 600,
+      item_tax_total: 112,
```
Comment (adrien2p) [inline packages/cart/integration-tests/__tests__/services/cart-module/index.spec.ts:2406]: q: would it be easier to display adjustment total for the item? just to make it easy to recompose the item_total

File: packages/cart/integration-tests/__tests__/services/cart-module/index.spec.ts [hunk 5]
```diff
@@ -2332,4 +2333,184 @@ describe("Cart Module Service", () => {
...
+
+      shipping_total: 116,
+      shipping_subtotal: 100,
+      shipping_tax_total: 16,
```
Comment (adrien2p) [inline packages/cart/integration-tests/__tests__/services/cart-module/index.spec.ts]: q: aren't we mnissing the adjustment here? I would have expected 96 for the shipping total no?

File: packages/cart/integration-tests/__tests__/services/cart-module/index.spec.ts [hunk 6]
```diff
@@ -2332,4 +2333,184 @@ describe("Cart Module Service", () => {
...
+
+      discount_total: 60,
+      discount_tax_total: 12,
+
+      original_total: 800,
```
Comment (adrien2p) [inline packages/cart/integration-tests/__tests__/services/cart-module/index.spec.ts:2415]: q: how did we get 800?

File: packages/cart/integration-tests/__tests__/services/cart-module/index.spec.ts [hunk 7]
```diff
@@ -2332,4 +2333,184 @@ describe("Cart Module Service", () => {
...
+          total: 480,
+          discount_total: 0,
+          tax_total: 80,
+        }),
+      ],
+      shipping_methods: [
+        expect.objectContaining({
+          amount: 100,
+          subtotal: 100,
+          total: 116, // (subtotal + tax - discount)
```
Comment (adrien2p) [inline packages/cart/integration-tests/__tests__/services/cart-module/index.spec.ts]: q: I believe the discount is not taken off here

File: packages/cart/src/services/cart-module.ts [hunk 8]
```diff
@@ -126,6 +130,54 @@ export default class CartModuleService<
...
+      "tax_total",
+      "shipping_total",
+      "discount_total",
+    ]
+
+    const shouldDecorate = (config?.select || []).some((field) =>
+      totalFields.includes(field)
+    )
+
+    // If no total fields are requested, we return early
```
Comment (adrien2p) [inline packages/cart/src/services/cart-module.ts]: todo: probably an old comment?

File: packages/cart/src/services/cart-module.ts [hunk 9]
```diff
@@ -126,6 +130,54 @@ export default class CartModuleService<
...
+          ...(config.relations || []),
+          "items.tax_lines",
+          "items.adjustments",
+          "shipping_methods.tax_lines",
+          "shipping_methods.adjustments",
+        ]),
+      }
+    }
+
+    const cart = await this.cartService_.retrieve(cartId, config, context)
```
Comment (adrien2p) [inline packages/cart/src/services/cart-module.ts:161]: q: shouldn't we remove the totals fields from the config select here?

File: packages/cart/src/services/cart-module.ts [hunk 10]
```diff
@@ -126,6 +130,54 @@ export default class CartModuleService<
...
+    return shouldDecorate ? decorateCartTotals(serialized) : serialized
+  }
+
+  @InjectManager("baseRepository_")
+  async retrieve(
+    cartId: string,
+    config: FindConfig<CartDTO>,
+    @MedusaContext() sharedContext: Context = {}
+  ): Promise<CartDTO> {
+    return await this.decorate(cartId, config, sharedContext)
```
Comment (adrien2p) [inline packages/cart/src/services/cart-module.ts:179]: q: do we also want that for the list*?

File: packages/utils/src/totals/shipping-method/index.ts [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+  } else {
+    const originalTotal = MathBN.add(
+      shippingMethod.amount,
+      totals.original_tax_total
+    )
+    const total = MathBN.add(shippingMethod.amount, totals.tax_total)
+    
+    totals.total = new BigNumber(total)
+    totals.original_total = new BigNumber(originalTotal)
+  }
```
Comment (adrien2p) [inline packages/utils/src/totals/shipping-method/index.ts]: todo: I believe we are missing the discount total here

Comment (olivermrbl) [reply packages/utils/src/totals/shipping-method/index.ts]: Yes, very nice catch. I didn't account for adjustments, as we haven't had that previously. 

Adding this should solve most of your other comments. Will get back on this.

File: packages/cart/integration-tests/__tests__/services/cart-module/index.spec.ts (hunk 2 above)
Comment (olivermrbl) [reply packages/cart/integration-tests/__tests__/services/cart-module/index.spec.ts:2337]: Will be removed once approach is approved.

File: packages/cart/src/services/cart-module.ts (hunk 10 above)
Comment (olivermrbl) [reply packages/cart/src/services/cart-module.ts:179]: Yup, will be added once the approach is approved.

File: packages/cart/integration-tests/__tests__/services/cart-module/index.spec.ts (hunk 6 above)
Comment (olivermrbl) [reply packages/cart/integration-tests/__tests__/services/cart-module/index.spec.ts:2415]:
```
original_total = (subtotal  + original_shipping_total + original_tax_total) - discount_total
original_total = (600  + 120 + 140) - 60
```

**Comment by olivermrbl** [comment]
Superceded by https://github.com/medusajs/medusa/pull/6980 and https://github.com/medusajs/medusa/pull/7010

**Pull request closed without merging**